File: python/Insertion Sort List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def insertionSortList(self, head):
        if head == None:
            return head
        i = head
        bSorted = True
        while i != None and i.next != None:
            if i.val > i.next.val:
                bSorted = False
                break
            i = i.next
        if bSorted is True:
            return head

        ret = ListNode(head.val)
        while head.next != None:
            head = head.next
            p = q = ret
            while p != None and p.val < head.val:
                q = p
                p = p.next
            newnode = ListNode(head.val)
            if p == q:
                newnode.next = p
                ret = newnode
            else:
                q.next = newnode
                newnode.next = p
        return ret

File: src/Insertion Sort List.cpp
#include <vector>
#include <iostream>
#include <stack>
#include <string>
#include <sstream>
using namespace std;


// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *insertionSortList(ListNode *head) {
        ListNode *retHead = new ListNode(0);
		ListNode *p, *q;
		if(!head)
			return head;
		while(head)
		{
			p = retHead;
			q = retHead->next;
			ListNode *pNode = new ListNode(head->val);
			while(q && q->val < head->val)
			{
				p = q;
				q = q->next;
			}
			p->next = pNode;
			pNode->next = q;
			head = head->next;
		}
		return retHead->next;
    }
};

int main()
{
	ListNode *node = new ListNode(1);
	Solution so;
	so.insertionSortList(node);
	return 0;
}